test/library/packages/Itertools/Itertools.chpl

```diff
@@ -105,4 +105,84 @@ module Itertools {
 
     for working_iters do yield arg;
   }
+
+
+
+  /*
+    Returns elements from an iterable over and over again, a specified
+    number of times.
+
+
+    :arg arg: The iterable whose elements are to be returned
+    :type arg: `?`
+
+    :arg times: The number of times to iterate through the iterable
+    (i.e. number of times each element is to be returned)
+    :type times: `int`
+
+    :yields: Elements of the iterable ``times`` times
+
+    :throws: ``IllegalArgumentError`` on parallel infinite iteration (see below)
+
+
+    If the argument ``times`` has the value 0, it will return each element of
+    the iterable an infinite number of times.
+
+    This iterator can be called in serial and parallel zippered contexts.
+
+    .. note::
+      This iterator is not suitable for parallel infinite iteration i.e.
+      avoid using zippered, ``forall``, or ``coforall`` loops with the
+      ``times`` argument set to 0.
+  */
+
+  // Serial iterator
+
+  iter cycle(arg, times = 0) {
+    if times == 0 then
+      for 0.. do
+        for element in arg do
+          yield element;
+    else
+      for 0..#times do
+        for element in arg do
+          yield element;
+  }
+
+  use RangeChunk;
+
+  // Parallel iterator - Leader
+
+  pragma "no doc"
+  iter cycle(param tag: iterKind, arg, times = 0) throws
+      where tag == iterKind.leader {
+
+    var numTasks = if dataParTasksPerLocale > 0 then dataParTasksPerLocale
+                                                else here.maxTaskPar;
+
+    if numTasks > times then numTasks = times;
+
+    if times == 0 then
+      throw new owned IllegalArgumentError(
+          "infinite iteration not supported for parallel loops");
+    else
+      coforall tid in 0..#numTasks {
+        const working_iters = chunk(0..#times, numTasks, tid);
+        yield(working_iters,);
+      }
+  }
+
+  // Parallel iterator - Follower
+
+  pragma "no doc"
+  iter cycle(param tag: iterKind, arg, times = 0, followThis)
+      where tag == iterKind.follower && followThis.size == 1 {
+
+    const working_iters = followThis(1);
+
+    for working_iters do
```

Comment:
> > A zippered iterations have non-equal lengths error is produced on running the same. However, when I try the same with a for/coforall loop, no error is produced (TIO), which I suppose means that there is some problem in the way the follower iterator receives the yielded range? How should I proceed with this?


This loop is actually doing `working_iters.size * arg.size` iterations rather than just `working_iters.size` iterations, which is why you get the non-equal lengths error.

You'll want to translate a given index into an element in the cycle in a consistent way, e.g.

```chpl
    for i in working_iters {
      yield arg[i%arg.size + 1];
    }
```

Here's a good test to ensure ordering is preserved:

```chpl
use Itertools;

var A: [1..10] string;
forall (id, el) in zip(1..9, cycle('ABC')) do A[id] = el;

writeln(A); // Should be A B C A B C A B C
```

However, this becomes tricky for iterables that cannot be indexed into such as domains and ranges. One way we could solve this is by creating overloads specific to arrays and strings (cases where we know we can index into them) with implementations like the example above, and keep a generic overload that create a temporary array of the cycled iterable so that we can index into it within the follower.

Here's a few tests you might try to get passing in your implementation:

```chpl
use Itertools;

var A: [1..10] int;
forall (id, el) in zip(1..10, cycle({1..4})) do A[id] = el;
writeln(A);
forall (id, el) in zip(1..10, cycle(1..4)) do A[id] = el;
writeln(A);
forall (id, el) in zip(1..10, cycle('ABC')) do A[id] = ascii(el):int;
writeln(A);
forall (id, el) in zip(1..10, cycle([1, 2, 3])) do A[id] = el;
writeln(A);

var B: [2..10] real;
forall (id, el) in zip(2..10, cycle({1..4})) do B[id] = el:real;
```

Reply:
> This loop is actually doing `working_iters.size * arg.size` iterations rather than just `working_iters.size` iterations, which is why you get the non-equal lengths error.

But for this tool, shouldn't the number of iterations be `working_iters.size * arg.size`?
For example, if one calls `cycle('ABC', 3)`, then the output should be `A B C A B C A B C`. If we do

> ```chapel
>     for i in working_iters {
>       yield arg[i%arg.size + 1];
>     }
> ```

then wouldn't it just print `A B C`?

Reply:
I think you mean that the total number of iterations should be `arg`*`times`, which is true. However, keep in mind that we're working in the follower iterator's scope here, where it will only provide a subset of the total iterations (`followThis(1) iterations).

The leader/follower API assumes that the follower iterator will only iterate `followThis(1)` times (assigned to `working_iters`), and so we need to ensure that we only iterate that number of times in a given follower iterator instance.